src/functional/fencil_processors/codegens/gtfn/itir_to_gtfn_ir.py

```diff
@@ -188,46 +249,148 @@ def visit_FunCall(self, node: itir.FunCall, **kwargs: Any) -> Node:
         elif isinstance(node.fun, itir.FunCall) and node.fun.fun == itir.SymRef(id="shift"):
             assert len(node.args) == 1
             return FunCall(
-                fun=self.visit(node.fun.fun), args=self.visit(node.args) + self.visit(node.fun.args)
+                fun=self.visit(node.fun.fun, **kwargs),
+                args=self.visit(node.args, **kwargs) + self.visit(node.fun.args, **kwargs),
             )
         elif isinstance(node.fun, itir.FunCall) and node.fun == itir.SymRef(id="shift"):
             raise ValueError("unapplied shift call not supported: {node}")
-        return FunCall(fun=self.visit(node.fun), args=self.visit(node.args))
+        return FunCall(fun=self.visit(node.fun, **kwargs), args=self.visit(node.args, **kwargs))
 
     def visit_FunctionDefinition(
         self, node: itir.FunctionDefinition, **kwargs: Any
     ) -> FunctionDefinition:
         return FunctionDefinition(
-            id=node.id, params=self.visit(node.params), expr=self.visit(node.expr)
+            id=node.id,
+            params=self.visit(node.params, **kwargs),
+            expr=self.visit(node.expr, **kwargs),
         )
 
+    @staticmethod
+    def _is_scan(node: itir.Node):
+        return isinstance(node, itir.FunCall) and node.fun == itir.SymRef(id="scan")
+
     def _visit_output_argument(self, node: itir.SymRef | itir.FunCall):
         if isinstance(node, itir.SymRef):
             return self.visit(node)
         elif isinstance(node, itir.FunCall) and node.fun == itir.SymRef(id="make_tuple"):
             return SidComposite(values=[self._visit_output_argument(v) for v in node.args])
         raise ValueError("Expected `SymRef` or `make_tuple` in output argument.")
 
-    def visit_StencilClosure(self, node: itir.StencilClosure, **kwargs: Any) -> StencilExecution:
-        assert isinstance(node.stencil, itir.SymRef)
-        backend = Backend(
-            domain=self.visit(node.domain, stencil=kwargs["symtable"][node.stencil.id], **kwargs)
-        )
+    @staticmethod
+    def _bool_from_literal(node: itir.Node):
+        assert isinstance(node, itir.Literal)
+        assert node.type == "bool" and node.value in ("True", "False")
+        return node.value == "True"
 
+    def visit_StencilClosure(
+        self, node: itir.StencilClosure, extracted_functions: list, **kwargs: Any
+    ) -> Union[ScanExecution, StencilExecution]:
+        backend = Backend(domain=self.visit(node.domain, stencil=node.stencil, **kwargs))
+        if self._is_scan(node.stencil):
+            scan_id = UIDs.sequential_id(prefix="_scan")
+            assert isinstance(node.stencil, itir.FunCall)
+            scan_lambda = self.visit(node.stencil.args[0], **kwargs)
+            forward = self._bool_from_literal(node.stencil.args[1])
+            scan_def = ScanPassDefinition(
+                id=scan_id, params=scan_lambda.params, expr=scan_lambda.expr, forward=forward
+            )
+            extracted_functions.append(scan_def)
```

Comment:
I am wondering if we should run this extraction as a separate pass on itir.

Reply:
Yes, this file clearly needs a huge cleanup. But not in this PR, because merging with existing code changes would probably be even more painful then.